src/trio/testing/_raises_group.py

```diff
@@ -545,50 +667,199 @@ def matches(
         if self.flatten_subgroups:
             actual_exceptions = self._unroll_exceptions(actual_exceptions)
 
-        # important to check the length *after* flattening subgroups
-        if len(actual_exceptions) != len(self.expected_exceptions):
+        self.fail_reason = _check_match(self.match_expr, exc_val)
+        if self.fail_reason is not None:
+            if (
+                len(actual_exceptions) == len(self.expected_exceptions) == 1
+                and isinstance(expected := self.expected_exceptions[0], type)
+                and isinstance(actual := actual_exceptions[0], expected)
+                and _check_match(self.match_expr, actual) is None
+            ):
+                assert (
+                    self.match_expr is not None
+                ), "can't be None if _check_match failed"
+                self.fail_reason += f", but matched expected exception. You might want RaisesGroup(Matcher({expected.__name__}, match={_match_pattern(self.match_expr)!r}"
             return False
 
-        for e in actual_exceptions:
-            for rem_e in remaining_exceptions:
-                if (
-                    (isinstance(rem_e, type) and isinstance(e, rem_e))
-                    or (isinstance(rem_e, RaisesGroup) and rem_e.matches(e))
-                    or (isinstance(rem_e, Matcher) and rem_e.matches(e))
-                ):
-                    remaining_exceptions.remove(rem_e)
+        # do the full check on expected exceptions
+        self.fail_reason = self._check_exceptions(
+            actual_exceptions,
+            _depth=_depth,
+        )
+
+        if self.fail_reason is not None:
+            # if we're not expecting a nested structure, and there is one, do a second
+            # pass where we try flattening it
+            if (
+                not self.flatten_subgroups
+                and not any(
+                    isinstance(e, RaisesGroup) for e in self.expected_exceptions
+                )
+                and any(isinstance(e, BaseExceptionGroup) for e in actual_exceptions)
+                and self._check_exceptions(
+                    self._unroll_exceptions(exc_val.exceptions),
+                    _depth=_depth,
+                )
+                is None
+            ):
+                self.fail_reason += (
+                    f"\n{_depth * '  '}Did you mean to use `flatten_subgroups=True`?"
+                )
+            return False
+
+        # only run `self.check` once we know `exc_val` is correct.
+        # _check_exceptions does not use TypeGuard, since we want to communicate the fail
+        # reason, so we need a type: ignore
+        # TODO: if this fails, we should say the *group* did not match
+        self.fail_reason = _check_check(self.check, exc_val, _depth)  # type: ignore[arg-type]
+
+        return self.fail_reason is None
+
+    def _check_exceptions(
+        self,
+        actual_exceptions: Sequence[BaseException],
+        _depth: int,
```

Comment:
I don't think this is necessary. I also don't like how it needs to be threaded everywhere. I've left further comments on where we recurse with a larger depth about indenting the result instead.